src/System.Security.SecureString/src/System/Security/SecureString.Windows.cs

```diff
@@ -0,0 +1,361 @@
+                }
+
+                if (bufferPtr != null)
+                {
+                    decryptedBuffer.ReleasePointer();
+                }
+            }
+            return result;
+        }
+
+        private void EnsureNotDisposed()
+        {
+            if (_encryptedBuffer == null)
+            {
+                throw new ObjectDisposedException(GetType().Name);
+            }
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private SafeBSTRHandle _encryptedBuffer;
+
+        private uint EncryptedBufferLength
+        {
+            get
+            {
+                Debug.Assert(_encryptedBuffer != null, "Buffer is not initialized!");
+                return _encryptedBuffer.Length;
+            }
+        }
+
+        private void AllocateBuffer(uint size)
+        {
+            _encryptedBuffer = SafeBSTRHandle.Allocate(size);
+        }
+
+        private void EnsureCapacity(ref SafeBSTRHandle decryptedBuffer, int capacity)
+        {
+            if (capacity > MaxLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_Capacity);
+            }
+
+            Debug.Assert(capacity > _decryptedLength, $"Expected to only be called when growing capacity = {capacity}, _decryptedLength = {_decryptedLength}");
+
+            SafeBSTRHandle newBuffer = SafeBSTRHandle.Allocate((uint)capacity);
+            SafeBSTRHandle.Copy(decryptedBuffer, newBuffer, (uint)_decryptedLength * sizeof(char));
+            decryptedBuffer.Dispose();
+            decryptedBuffer = newBuffer;
+        }
+
+        private void ProtectMemory(SafeBSTRHandle decryptedBuffer)
+        {
+            Debug.Assert(!decryptedBuffer.IsInvalid, "Invalid buffer!");
+
+            if (_decryptedLength == 0)
+            {
+                return;
+            }
+
+            try
+            {
+                SafeBSTRHandle newEncryptedBuffer;
+                if (!CryptProtectData(decryptedBuffer, out newEncryptedBuffer))
+                {
+                    throw new CryptographicException(Marshal.GetLastWin32Error());
+                }
+
+                _encryptedBuffer.Dispose();
+                _encryptedBuffer = newEncryptedBuffer;
+            }
+            finally
+            {
+                decryptedBuffer.ClearBuffer();
+                decryptedBuffer.Dispose();
+            }
+        }
+
+        private SafeBSTRHandle UnprotectMemory()
+        {
+            Debug.Assert(!_encryptedBuffer.IsInvalid, "Invalid buffer!");
+
+            if (_decryptedLength == 0)
+            {
+                return _encryptedBuffer;
+            }
+
+            SafeBSTRHandle decryptedBuffer;
+            if (!CryptUnprotectData(_encryptedBuffer, out decryptedBuffer))
+            {
+                throw new CryptographicException(Marshal.GetLastWin32Error());
```

Comment:
Due to the cross-platforminess we don't automatically look up error strings in this constructor.